packages/core/src/node/cluster/main.ts

```diff
@@ -36,6 +36,10 @@ export interface Address {
 export async function start(serverPath: string): Promise<Address> {
     if (isMaster) {
         const master = new MasterProcess();
+        master.on('exit', () =>
+            Object.keys(cluster.workers).length === 0
```

Comment:
checking workers should not be necessary if you do https://github.com/theia-ide/theia/pull/1893#discussion_r189159070